### PR TITLE
Studio: NumberFormatParsing: Deal with systems with split

### LINKED ISSUE #2437: Locale-dependent parsing of Pixel Size (um) gives wrong affine matrix
When opening the Pixel preset editor, `Pixel Size (um)` is initiated with `0.00`. When entering a value in the same format, e.g. `1.30` and then click on `Calculate`, the affine matrix gets populated with `130,00000`. When `1,30` is entered (comma instead of point), affine gets populated correctly with `1,30000`.

Could be a locale issue, where the `.` is being read as a thousands separator instead of a decimal point, so 1.30 becomes 130. Seems likely that other numeric fields are also affected. 

Initial state:
Image
Entering with `.` separator:
Image
`1.3` is parsed as `130`:
Image
`1,3` gets parsed correctly:
Image

### PR DESCRIPTION
**Pull request by nicost**
locals, such as en-ch.  Needs testing on such systems. Closes #2437

### CODE REVIEW

File: mmstudio/src/test/java/org/micromanager/internal/utils/NumberUtilsTest.java [hunk 1]
```diff
@@ -88,4 +88,34 @@ public void displayToDoubleIsCorrect() throws java.text.ParseException {
...
+    * A correctly-grouped number in the same locale must still parse correctly.
+    */
+   @Test
+   public void displayStringToDoubleRejectsInvalidGrouping() throws java.text.ParseException {
+      java.text.DecimalFormatSymbols symbols =
+            new java.text.DecimalFormatSymbols(java.util.Locale.getDefault());
+      char group = symbols.getGroupingSeparator();
+      char decimal = symbols.getDecimalSeparator();
+      org.junit.Assume.assumeTrue(group != decimal);
+
```
[the remark this answers is not in the document]
Comment (nicost) [reply mmstudio/src/test/java/org/micromanager/internal/utils/NumberUtilsTest.java]: Done.

File: mmstudio/src/main/java/org/micromanager/internal/utils/NumberUtils.java [hunk 2]
```diff
@@ -84,15 +86,54 @@ public static String doubleToDisplayString(double number, int maxPrecision) {
...
+   private static Number parseStrict(String numberString) throws ParseException {
+      if (FORMAT instanceof DecimalFormat) {
+         DecimalFormatSymbols symbols = ((DecimalFormat) FORMAT).getDecimalFormatSymbols();
+         char groupingSep = symbols.getGroupingSeparator();
+         char decimalSep = symbols.getDecimalSeparator();
+         if (groupingSep != decimalSep && numberString.indexOf(groupingSep) >= 0) {
+            String group = Pattern.quote(String.valueOf(groupingSep));
+            String decimal = Pattern.quote(String.valueOf(decimalSep));
+            String validNumber = "[+-]?\\d{1,3}(" + group + "\\d{3})*(" + decimal + "\\d+)?";
+            if (!numberString.matches(validNumber)) {
```
[the remark this answers is not in the document]
Comment (nicost) [reply mmstudio/src/main/java/org/micromanager/internal/utils/NumberUtils.java]: Addressed

**Comment by hinderling** [comment]
Tested on the same machine from #2437 (Windows UI locale `en-US`, regional format `en-CH`, so Java sees DISPLAY=`en_US` / FORMAT=`en_CH` -> decimal `,`, grouping `.`).

Quick table with different input values;' values are what `NumberUtils.displayStringToDouble()` actually returns. `error` means a `ParseException`, which the Pixel preset editor shows as "Pixel Size is not a valid Number".

| typed   | expected | main | PR #2444 |
|---------|----------|------|----------|
| `1.3`   | 1.3      | 13   | error    |
| `1.30`  | 1.3      | 130  | error    |
| `1.300` | 1.3      | 1300 | 1300     |
| `1,3`   | 1.3 / error    | 1.3  | 1.3      |
| `0.00`  | 0        | 0    | error    |

What is still confusing to me is that numbers formated like the template (0.00) get refused, and the difference between valid/invalid groupings (`1.30` vs`1.300`): the patch catches one and two decimals, but three decimals after the `.` look like a valid grouping (which seems to be a common case for pixel sizes like `0.065`,).

And fully opinionated (but think it might reflect general DE/CH population); since I never type grouping separators, I'd expect a `.` to always be read as a decimal separator regardless of my locale, and a `,` to either be rejected or treated as a `.` (that's what I'm used to from other software). More generally, I'm not sure if the effort is worth to adjust number formatting to locale?

Two more notes:
- Seems it actually isn't specifically about split locale? A plain `de-DE` user typing `0.325` also would get 325 on both `main` and the PR, and a plain `en-US` user typing `0,325` gets 325. (Split locale just made it easy to trip over)
- As `0.00` is the literal that `CalibrationListDlg.java:304` seeds a new calibration with, opening "New" and clicking OK without touching the field now errors.

**Comment by nicost** [comment]
Thanks!  Can you try again with the latest additions?  Sorry to spring this on you, but you seem much better equipped to test locales...

**Comment by hinderling** [comment]
Typed input now does exactly what I'd hoped, and new `NumberUtilsLocaleTest` passes here:

| typed   | expected | now  |
|---------|----------|------|
| `1.3`   | 1.3      | 1.3  |
| `1.30`  | 1.3      | 1.3  |
| `1.300` | 1.3      | 1.3  |
| `1,3`   | 1.3      | 1.3  |
| `0.00`  | 0        | 0    |

But `doubleToDisplayString()` still formats with the locale *and with grouping*, while the new parser reads `.` and `,` as decimal separators. So the two no longer agree, and values the app displays itself no longer survive a round trip:

| value    | displayed as | read back | |
|----------|--------------|-----------|---|
| `1.3`    | `1,3`        | 1.3       | ok |
| `999.5`  | `999,5`      | 999.5     | ok |
| `1000`   | `1.000`      | **1.0**   |x |
| `1234.5` | `1.234,5`    | **error** |x |
| `250000` | `250.000`    | **250.0** | x |

On `main` all six of these round-trip correctly, so this is new. It's the same failure mode as #2437, just from the other direction (this time string was produced by Micro-Manager, not mistyped by me). Anything >= 1000 in a field that gets populated from a double and read back is affected (stage coordinates in um, exposure times, z ranges).

Adding one line to the static block should fix these:

```java
FORMAT.setGroupingUsed(false);
```

which should make the formatter stop emitting a separator the parser will not accept (`1000` -> `"1000"`, `1234.5` -> `"1234,5"`). 

Three smaller things:

- `displayStringToInt()` and `displayStringToLong()` still use the old `FORMAT.parse()`, so the original bug is still live there: `1.3` -> 13, `1.30` -> 130, `1.300` -> 1300, and `1,3` -> 1. That path is used for e.g. the buffer size in `OptionsDlg` and the autofocus skip interval in `AcqControlDlg`. Doubles and ints now also disagree with each other.
- `Double.parseDouble()` is more permissive than `NumberFormat` was, and now accepts `NaN` -> NaN, `Infinity` -> Infinity, `1d`/`2F` -> 1.0/2.0 (Java type suffixes) and `0x1p3` -> 8.0. NaN/Infinity reaching `setPixelSizeUm()` seems worth blocking.
- `displayStringToDouble(null)` now throws `NullPointerException` rather than `ParseException`, so callers that catch `ParseException` (e.g. `getPixelSize()`) would not catch it.

**Comment by nicost** [comment]
I believe I addressed all of those (great!) points.  Would love it if you can test again.

**Comment by hinderling** [comment]
Great!! Everything as expected now.

few small leftovers:
- `NumberUtilsTest` still fails for me: 1 of 5, `doubleToStringIsCorrect`, `expected:<0[.]0001> but was:<0[,]0001>`. Pre-existing `@Before` guard, which skips based on `java.util.Locale.getDefault()`, but the thing it is guarding (`doubleToDisplayString` -> `FORMAT` -> `NumberFormat.getInstance()`) follows `Locale.getDefault(Locale.Category.FORMAT)`. On a split-locale machine those disagree, so the guard passes and then the test fails. Changing that one line to `Locale.getDefault(java.util.Locale.Category.FORMAT)` gives `OK (5 tests)` here.
- Very much an edge case; `1e400` still gives `Infinity` and `1e-400` gives `0.0`, since the pattern allows an exponent and `Double.parseDouble` overflows silently. As literal `Infinity` is blocked maybe you want to block this infinity too? Underflow -> 0 seems reasonable to keep.
- one cosmetic thing: a leading `+` is accepted as a double but rejected as an int, so the same string parses in one field and not the other:

| typed  | as double | as int |
|--------|-----------|--------|
| `3`    | 3.0       | 3      |
| `-3`   | -3.0      | -3     |
| `+3`   | 3.0       | error (wrong?) |
| `3.1`  | 3.1       | error  (correct) |
| `+3.1` | 3.1       | error  (correct) |

odd one out seems`+3`. It's only `-?` vs `[+-]?`: `DECIMAL_NUMBER_PATTERN` accepts either sign, while `validNumber` in `parseIntegerStrict` accepts only `-`. Would need changing in both branches (`-?\d+` and the grouped alternative), otherwise `+10.000` would still fail while `+10000` passes.

**Comment by hinderling** [comment]
if you think those are worth patching -- i made a small PR that addresses these https://github.com/nicost/micro-manager/pull/104

**Comment by nicost** [comment]
Thank you so much!  I believe this is ready to be merged now.  Hopefully this is the end of locale issues!
